### DISCUSSION TITLE
Enable loading migrations from multiple directories.

### DISCUSSION THREAD

**Discussion by pedestrianlove** [SeaQL/sea-orm#2189] [Q&A]
I'm new to Rust, so I may have missed some obvious hints. If that's the case, please pardon my ignorance. 🙏

---

I'm looking for a way to implement modular design in my project using SeaORM. 

e.g. Each module folder may correspond to a distinct set of features in the system, and they have their migrations, models, controllers, and routes.

However, I had a difficult time finding a way to put migrations into separate module folders. (at least I cannot find it in [docs](https://www.sea-ql.org/SeaORM/docs/index/))

#### Is it possible to achieve this kind of behavior in SeaORM?

Thanks!

**Comment by pedestrianlove** [comment]
After looking into the source of the scaffolded crate, I just found that you can customize the behavior of `sea-orm-cli migrate` by modifying the content of `main.rs`.  In theory, you should be able to stack multiple customized `Migrator` instances so that they're all run within the same `main.rs`, triggered by the same sea-orm-cli command.

Therefore, I'm changing the category of this discussion from Ideas to Q&A for now to avoid possible confusion. (I'm going to try if this approach works or not)

(btw, is this a recommended practice to do so, or does SeaORM have a requirement or limitation that depends on that `cli` line of code to be run only once?)

**Comment by pedestrianlove** [comment]
#### tl;dr: The "Multiple migrator" approach either doesn't work or requires additional modification.

The following is my attempt making use of `sea_orm_cli` crate to migrate programmatically:

---
`migrations/src/main.rs`
```rust
use sea_orm_migration::prelude::*;

#[async_std::main]
async fn main() {
    cli::run_cli(migration::KernelMigrator).await;
    cli::run_cli(migration::NotesMigrator).await;
}
```
`migrations/src/lib.rs`
```rust
#![allow(elided_lifetimes_in_paths)]
#![allow(clippy::wildcard_imports)]
pub use sea_orm_migration::prelude::*;

mod m20220101_000001_users;
mod m20231103_114510_notes;

pub struct KernelMigrator;
pub struct NotesMigrator;


#[async_trait::async_trait]
impl MigratorTrait for KernelMigrator {
    fn migrations() -> Vec<Box<dyn MigrationTrait>> {
        vec![
            Box::new(m20220101_000001_users::Migration),
        ]
    }
}

#[async_trait::async_trait]
impl MigratorTrait for NotesMigrator {
    fn migrations() -> Vec<Box<dyn MigrationTrait>> {
        vec![
            Box::new(m20231103_114510_notes::Migration),
        ]
    }
}
```

`Result:`
```shell
➜  tester $ DATABASE_URL="mysql://test:test@localhost:3306/tester" sea-orm-cli migrate up                                                                                                                                                                                                                                                                                                                                                                                         (base) 
Running `cargo run --manifest-path ./migration/Cargo.toml -- up -u mysql://test:test@localhost:3306/tester`
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/migration up -u 'mysql://test:test@localhost:3306/tester'`
Applying all pending migrations
Applying migration 'm20220101_000001_users'
Migration 'm20220101_000001_users' has been applied
thread 'main' panicked at /home/jsl/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-subscriber-0.3.18/src/util.rs:91:14:
failed to set global default subscriber: SetGlobalDefaultError("a global default trace dispatcher has already been set")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Fail to run migration
```

**Comment by lucas-labs** [comment]
Did you find a workaround? I'm facing the same issue trying to call `cli::run_cli` multiple times with different migrators. In my case, I have several libraries and each has their own migrators creating the tables they need. Then I have a "main" project which generates the final binary. My idea was to call `run_cli` for each migrator in the main, but... nope.

**Reply by pedestrianlove** [reply]
Not yet. And our team just decided to move on to golang instead, so we're unlikely to actively explore this issue in near future.

**Comment by lucas-labs** [comment]
Btw this is what I'm doing for now, unless there's a better way (create a "`Main`" migrator, that combines the migrations of all the other migrators:

```rust
use sea_orm_migration::prelude::*;

/// main migrator, which combines all the other migrations
pub struct MainMigrator;

#[async_trait::async_trait]
impl MigratorTrait for MainMigrator {
    fn migrations() -> Vec<Box<dyn MigrationTrait>> {
        let mut migrations = vec![];

        // get migrations from the specific migrators
        migrations.extend(lib_foo::migration::Migrator::migrations());
        migrations.extend(lib_bar::migration::Migrator::migrations());
        .... add as many migrators as we want ...
  
        // return the combined migrations vector
        migrations
    }
}

#[async_std::main]
async fn main() {
    // execute em!
    cli::run_cli(MainMigrator).await;
}
```
